### PR TITLE
feat: Link preview settings to Status chat input unfurling mode

### PR DESCRIPTION
**Pull request by alexjba**
### What does the PR do

Closing #12442

Changes:
1. Update privacy controller to react to external unfurling settings change
2. Add handler to input controller and react to external unfurling settings change
3. Small other fixes

### Affected areas

Link preview settings

#### Changes in settings visible in chat input

https://github.com/status-im/status-desktop/assets/47811206/c47a3b9a-9fd3-46c9-9d61-e588721e0ba2

#### Changes in chat input visible in settings

https://github.com/status-im/status-desktop/assets/47811206/e647b9bd-c229-4b72-b5c9-a706b6a579af

- [x] I've checked the design and this PR matches it

### CODE REVIEW

**Review by caybro — APPROVED** [review]
Nice, looks good to me!

File: src/app/modules/main/profile_section/privacy/view.nim [hunk 1]
```diff
@@ -66,8 +66,8 @@ QtObject:
   proc setUrlUnfurlingMode(self: View, value: int) {.slot.} =
     if self.getUrlUnfurlingMode() == value:
       return
-    self.delegate.setUrlUnfurlingMode(value)
-    self.urlUnfurlingModeChanged()
+    if self.delegate.setUrlUnfurlingMode(value):
+      self.urlUnfurlingModeChanged()
```
Comment (igor-sirotin) [inline src/app/modules/main/profile_section/privacy/view.nim:70]: It seems that we have a double-signalling here now.
Previous line will lead to this call-stack, calling same signal at the end.

1. `setUrlUnfurlingMode` 
2. `settings_service.saveUrlUnfurlingMode` 
3. `emit(SIGNAL_URL_UNFURLING_MODE_UPDATED)` 
4. `onUrlUnfurlingModeUpdated()`
5. `urlUnfurlingModeChanged()`

Comment (igor-sirotin) [reply src/app/modules/main/profile_section/privacy/view.nim:70]: So we probably don't need to emit `urlUnfurlingModeChanged` here at all. 
And then `setUrlUnfurlingMode` won't need to return `bool` result.

File: ui/app/AppLayouts/Chat/views/ChatColumnView.qml [hunk 2]
```diff
@@ -167,14 +165,6 @@ Item {
         }
     }
 
-    Connections {
```
Comment (igor-sirotin) [inline ui/app/AppLayouts/Chat/views/ChatColumnView.qml:170]: I've fixed this connection in my merged PR, but your solution is more correct, we don't need to check `DisableAll` here. Just a note for you for easier rebase, simply take your code here 🙂

File: src/app/modules/main/chat_section/chat_content/input_area/controller.nim [hunk 3]
```diff
@@ -229,11 +234,16 @@ proc loadLinkPreviews*(self: Controller, urls: seq[string]) =
...
-  else:
+  elif not enabled and self.settingsService.saveUrlUnfurlingMode(UrlUnfurlingMode.Disabled):
     self.linkPreviewPersistentSetting = UrlUnfurlingMode.Disabled
     self.linkPreviewCurrentMessageSetting = UrlUnfurlingMode.Disabled
 
   self.delegate.setAskToEnableLinkPreview(false)
+
+proc onUnfurlingModeChanged(self: Controller, value: UrlUnfurlingMode) =
+  self.linkPreviewPersistentSetting = value
+  self.resetLinkPreviews()
```
Comment (igor-sirotin) [inline src/app/modules/main/chat_section/chat_content/input_area/controller.nim:248]: Hmmm, I don't get it 🤔 
`resetLinkPreviews` will set local setting to `AlwaysAsk`:
https://github.com/status-im/status-desktop/blob/f2a20631d60808c2c4e56db878d572d4680b9974/src/app/modules/main/chat_section/chat_content/input_area/controller.nim#L116

Though initially you sync those variables values:
https://github.com/status-im/status-desktop/blob/f2a20631d60808c2c4e56db878d572d4680b9974/src/app/modules/main/chat_section/chat_content/input_area/controller.nim#L56-L57

File: src/app/modules/main/chat_section/chat_content/input_area/controller.nim [hunk 4]
```diff
@@ -229,11 +234,16 @@ proc loadLinkPreviews*(self: Controller, urls: seq[string]) =
...
+  self.setText(self.delegate.getText(), self.linkPreviewPersistentSetting == UrlUnfurlingMode.Enabled)
```
Comment (igor-sirotin) [inline src/app/modules/main/chat_section/chat_content/input_area/controller.nim:249]: Should we use `self.getLinkPreviewEnabled()` instead? 
It seems that this won't work for `AlwaysAsk` persistent setting.

**Review by igor-sirotin — COMMENTED** [review]
Thanks for finalising the settings! ❤️ 

1. I tested and it works like a magic! 🎉 
2. I also tested this with paired accounts, works as well:

https://github.com/status-im/status-desktop/assets/25482501/1a60357a-2939-4bed-8b1d-e0a1d805d080

### Minor concerns though

I have some concerns, not sure if those needs to be fixed in this PR:

1. Status links should always be unfurled.
Even when the mode is `Never show link previews`, status links should be unfurled, here's the corresponding status-go PR: https://github.com/status-im/status-go/pull/4159. This PR points to a status-go version that includes required changes.

Screenshot 2023-10-18 at 12 30 03
Screenshot 2023-10-18 at 12 30 06

2. When unfurling mode is switched from `Always ask` to `Enabled` an extra preview card quickly appears in the chat input.
It's not possible to see on a single instance, but's easy to see with 2 instances.
I don't say that it's important to fix 😄 But it might be a a symptom of that's smth wrong with implementation.

https://github.com/status-im/status-desktop/assets/25482501/7ee7bbc3-b92c-4688-b2dd-cc24ae05c124

File: src/app/modules/main/chat_section/chat_content/input_area/controller.nim (hunk 3 above)
Comment (alexjba) [reply src/app/modules/main/chat_section/chat_content/input_area/controller.nim:249]: ohh, you're right! On reset is should pick the persisted value.
Fixed

File: src/app/modules/main/profile_section/privacy/view.nim (hunk 1 above)
Comment (alexjba) [reply src/app/modules/main/profile_section/privacy/view.nim]: Good catch! Thanks!
Fixed

**Review by noeliaSD — APPROVED** [review]
Tested! Code LGTM!

File: src/app/modules/main/profile_section/privacy/view.nim [hunk 5]
```diff
@@ -61,10 +61,8 @@ QtObject:
   proc getUrlUnfurlingMode(self: View): int {.slot.} =
     return self.delegate.urlUnfurlingMode()
   proc setUrlUnfurlingMode(self: View, value: int) {.slot.} =
-    if self.getUrlUnfurlingMode() == value:
-      return
```
Comment (igor-sirotin) [inline src/app/modules/main/profile_section/privacy/view.nim:65]: hmm, why not?

**Review by igor-sirotin — APPROVED** [review]
looks good, thanks for fixes!

File: src/app/modules/main/profile_section/privacy/view.nim (hunk 5 above)
Comment (alexjba) [reply src/app/modules/main/profile_section/privacy/view.nim]: Fixing a non-issue and introducing a real one.. Thanks for spotting it! Removed

**Pull request merged**